### PR TITLE
Modifies util::load_private_key to support PKCS8, adds tests

### PR DESCRIPTION
**Pull request by peterdelevoryas**
Currently, `util::load_private_key` does not handle the modern private key format, `PKCS8`. `rustls` provides a utility for reading `PKCS8` formatted private keys, so this pull request just modifies `load_private_key` to attempt reading one `PKCS8` key, or else one `RSA` key. This is the behavior used in the `rustls` [tls server](https://github.com/ctz/rustls/blob/master/examples/tlsserver.rs) example.

This commit also includes two private key files, one formatted in the traditional `RSA` format, one in the `PKCS8` format, and adds tests for loading each one.

I'm pretty sure this should fix whatever issues people were having at the end of the `rocket` SSL support issue: SergioBenitez/Rocket#28

Things that I wasn't sure about/you might want me to change:

- I didn't increment the version number
- I made two tests, one for `PKCS8` and one for `RSA`, instead of a single test for both
- If it fails to load a `PKCS8` key, I seek to the start of the file and try `RSA`, instead of reopening the file
- I used combinator functions instead of match and if/else expressions
- I put the key files used in the tests in the `tests` directory 

Edit: When I made this, I wasn't entirely sure what the `RSA` format was, I was mostly just trying to be consistent with the naming in `rustls`. After some reading, I now realize that the function I used is for loading `PKCS1` `PEM` private keys. I thought I would just clarify this, because it nonsensical to call one format `RSA` and the other `PKCS8`, because they can both be `RSA` private keys. I also realized that `PKCS8` can contain other types of private keys (`DSA` or `DH`). I couldn't actually find an `openssl` command to convert a `DH PARAMETERS` `PEM` file into a `PKCS8` file, so I couldn't verify this, but it would likely give users a hard error (because of the `RSASigner::new` check). Obviously I'm not an expert on these things, so it would be great if someone more knowledgeable could comment on this.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -266,18 +266,31 @@ pub mod util {
...
+            }
         }
 
+        let keyfile = fs::File::open(filename).map_err(Error::Io)?;
+        let mut reader = BufReader::new(keyfile);
+        // Try reading one pkcs8 private key, or else one rsa private key
+        let key = pemfile::pkcs8_private_keys(&mut reader)
+            .map_err(|_| Error::BadKey)
+            .and_then(expect_one)
+            .or_else(|_| {
```
Comment (SergioBenitez) [inline src/lib.rs]: This `or_else` will called if reading/parsing a PKCS#8 file succeeds but `expect_one` fails. We probably don't want to try to read the file again in this case. Further, the user will get the wrong error message.

File: src/lib.rs [hunk 2]
```diff
@@ -266,18 +266,31 @@ pub mod util {
...
-        let mut keys = pemfile::rsa_private_keys(&mut reader)
-            .map_err(|_| Error::BadKey)?;
-
-        // Ensure there's only one key in the file.
-        if keys.len() != 1 {
-            return Err(Error::BadKeyCount);
+        use std::io::Seek;
+        let start = io::SeekFrom::Start(0);
+
+        fn expect_one(mut keys: Vec<rustls::PrivateKey>) -> Result<rustls::PrivateKey> {
```
Comment (SergioBenitez) [inline src/lib.rs]: `expect` usually implies that a `panic!` might take place. Let's rename this to `take_one` or something of that sort.

File: src/lib.rs [hunk 3]
```diff
@@ -266,18 +266,31 @@ pub mod util {
...
-        let keyfile = fs::File::open(filename).map_err(|e| Error::Io(e))?;
-        let mut reader = BufReader::new(keyfile);
-        let mut keys = pemfile::rsa_private_keys(&mut reader)
-            .map_err(|_| Error::BadKey)?;
-
-        // Ensure there's only one key in the file.
-        if keys.len() != 1 {
-            return Err(Error::BadKeyCount);
+        use std::io::Seek;
+        let start = io::SeekFrom::Start(0);
```
Comment (SergioBenitez) [inline src/lib.rs]: We only use this variable once, so we can probably just inline the `io::SeekFrom::Start(0)` at that location.

**Review by SergioBenitez — CHANGES_REQUESTED** [review]
This is great. We should definitely accept both PKCS#1 (`BEGIN RSA PRIVATE KEY`) and PKCS#8 (`BEGIN PRIVATE KEY`) PEM files. A few changes and we'll get this in.

File: src/lib.rs [hunk 4]
```diff
@@ -266,18 +266,33 @@ pub mod util {
...
+        let keyfile = fs::File::open(filename).map_err(Error::Io)?;
         let mut reader = BufReader::new(keyfile);
-        let mut keys = pemfile::rsa_private_keys(&mut reader)
-            .map_err(|_| Error::BadKey)?;
 
-        // Ensure there's only one key in the file.
-        if keys.len() != 1 {
-            return Err(Error::BadKeyCount);
-        }
+        let private_keys_fn = {
```
Comment (SergioBenitez) [inline src/lib.rs]: Can you add a comment about why this is happening here, for future reference? Just something short and pithy, nor more than a couple of lines.

File: src/lib.rs [hunk 5]
```diff
@@ -266,18 +266,33 @@ pub mod util {
...
+            let mut line = String::new();
+            reader.read_line(&mut line).map_err(Error::Io)?;
+            let private_keys_fn = match line.trim_right() {
+                "-----BEGIN RSA PRIVATE KEY-----" => pemfile::rsa_private_keys,
+                "-----BEGIN PRIVATE KEY-----" => pemfile::pkcs8_private_keys,
+                _ => return Err(Error::BadKey)
+            };
+            reader.seek(io::SeekFrom::Start(0)).map_err(Error::Io)?;
```
Comment (SergioBenitez) [inline src/lib.rs]: Let's add an empty line above here.

**Review by SergioBenitez — CHANGES_REQUESTED** [review]
This looks great! Just a couple of stylistic changes. :)

File: src/lib.rs (hunk 5 above)
Comment (peterdelevoryas) [reply src/lib.rs]: I actually went a couple steps further on that part, making another nested block and making the match statement the final expression for `private_keys_fn`. Sorry about that, if it's too much I can change it back to what it was.

Comment (SergioBenitez) [reply src/lib.rs]: Why do we need any nested blocks at all? Can we just do:

```rust
let mut first_line = String::new();
reader.read_line(&mut line).map_err(Error::Io)?;
reader.seek(io::SeekFrom::Start(0)).map_err(Error::Io)?;

let private_keys_fn = match first_line.trim_right() {
    "-----BEGIN RSA PRIVATE KEY-----" => pemfile::rsa_private_keys,
    "-----BEGIN PRIVATE KEY-----" => pemfile::pkcs8_private_keys,
    _ => return Err(Error::BadKey)
};
```

Comment (peterdelevoryas) [reply src/lib.rs]: No reason at all, that is much simpler! I'm going to amend the revision.

(In the past, I have tended to use nested blocks frequently to isolate the scope and mutability of variables from one another, but I get the feeling that I might be creating an anti-pattern that makes simple code more complex/harder-to-maintain/harder-to-reader/etc)

**Review by SergioBenitez — APPROVED** [review]

**Pull request merged**